File: 77. Combinations.java
/*
Problem Link: https://leetcode.com/problems/combinations/
*/

import java.util.*;

class Solution {
    private void helper(List<List<Integer>> result, List<Integer> temp, int n, int k, int start) {
        if (temp.size() == k) {
            result.add(new ArrayList<>(temp));
            return;
        }
        for (int i = start; i <= n; i++) {
            temp.add(i);
            helper(result, temp, n, k, i + 1);
            temp.remove(temp.size() - 1);
        }
    }

    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        helper(result, temp, n, k, 1);
        return result;
    }
}

File: Problems/77. Combinations.py
"""
Problem Link: https://leetcode.com/problems/combinations/
"""


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        if (n == 0 or k == 0):
            return []

        result = []
        for i in range(1, n + 1):
            result.append([i])

        for i in range(1, k):
            temp = []
            for j in range(len(result)):
                for k in range(result[j][-1] + 1, n + 1):
                    temp.append(result[j] + [k])
            result = temp

        return result
